File: Practicas/Qytetet/QytetetRuby/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "especulador"

module ModeloQytetet
  class Jugador
    attr_reader :nombre, :saldo, :propiedades, :fianza
    attr_accessor :carta_libertad, :casilla_actual, :encarcelado, :tiene_carta_libertad
    
    def initialize(e, n, s, tcl, cl, p, ca)
      @encarcelado = e
      @nombre = n
      @saldo = s
      @tiene_carta_libertad = tcl
      @carta_libertad = cl
      @propiedades = p
      @casilla_actual = ca
    end
    
    def self.nuevo(n,s)
      self.new(false, n, s, false, nil, Array.new, 0)
    end
    
    def self.copia(j)
      self.new(j.encarcelado, j.nombre, j.saldo, j.tiene_carta_libertad,
               j.carta_libertad, j.propiedades, j.casilla_actual)
    end

    def <=>(otro_jugador)            
      otro_jugador.obtener_capital <=> obtener_capital     
    end
    
    def cancelar_hipoteca(titulo)
      coste_cancelar_hipoteca = titulo.calcular_coste_cancelar
      cancelada = false
      if(tengo_saldo(coste_cancelar_hipoteca) && es_de_mi_propiedad_hipotecada(titulo,true))
          modificar_saldo(-titulo.calcular_coste_cancelar)
          titulo.cancelar_hipoteca
          cancelada = true
      end
      cancelada
    end

    def comprar_titulo_propiedad
      comprado = false
      coste_compra = @casilla_actual.coste
      if(tengo_saldo(coste_compra))
        titulo = @casilla_actual.asignar_propietario(self)
        comprado = true
        @propiedades << titulo
        modificar_saldo(-coste_compra)
      end
      comprado
    end

    def convertirme(fianza)
      esp = Especulador.copia(self,fianza)
      esp
    end
    
    def cuantas_casas_hoteles_tengo
      contador = 0
      @propiedades.each do |p|
          contador = contador + p.num_casas + p.num_hoteles
      end
      contador
    end

    def debo_ir_a_carcel
      !@tiene_carta_libertad
    end
    
    def debo_pagar_alquiler
      es_de_mi_propiedad = es_de_mi_propiedad(@casilla_actual.titulo)
      tiene_propietario = false
      esta_encarcelado = false
      esta_hipotecada = false
      if(!es_de_mi_propiedad)
          tiene_propietario = @casilla_actual.tengo_propietario
          if(tiene_propietario)
              esta_encarcelado = @casilla_actual.propietario_encarcelado
              esta_hipotecada = @casilla_actual.titulo.hipotecada
          end
      end
      !es_de_mi_propiedad & tiene_propietario & !esta_encarcelado & !esta_hipotecada
    end

    def devolver_carta_libertad
      @tiene_carta_libertad = false
      @carta_libertad
    end

    def edificar_casa(titulo)
      edificada = false
      if(puedo_edificar_casa(titulo))
        coste_edificar_casa = titulo.precio_edificar
        tengo_saldo = tengo_saldo(coste_edificar_casa)
        if(tengo_saldo)
            titulo.edificar_casa
            modificar_saldo(-coste_edificar_casa)
            edificada = true
        end
      end
      edificada
    end

    def edificar_hotel(titulo)
      edificada = false
      if(puedo_edificar_hotel(titulo))
        coste_edificar_hotel = titulo.precio_edificar
        tengo_saldo = tengo_saldo(coste_edificar_hotel)
        if(tengo_saldo)
            titulo.edificar_hotel
            modificar_saldo(-coste_edificar_hotel)
            edificada = true
        end
      end
      edificada
    end

    def eliminar_de_mis_propiedades(titulo)
      @propiedades.delete(titulo)
      titulo.propietario = nil
    end

    def es_de_mi_propiedad(titulo)
      resultado = false
      @propiedades.each do |p|
        if(p == titulo)
            resultado = true
        end
      end
      resultado
    end

    def es_de_mi_propiedad_hipotecada(titulo, hipotecada)
      resultado = false
      propiedades_hipotecadas = obtener_propiedades(hipotecada)
      propiedades_hipotecadas.each do |p|
        if(p == titulo)
            resultado = true
        end
      end
      resultado
    end
=begin
    def estoy_en_calle_libre

    end
=end
    def hipotecar_propiedad(titulo)
      hipotecada = false
      if(es_de_mi_propiedad_hipotecada(titulo,false))
        coste_hipoteca = titulo.hipotecar
        modificar_saldo(coste_hipoteca)
        hipotecada = true
      end
      hipotecada
    end
    
    def ir_a_carcel(casilla)
      @casilla_actual = casilla
      @encarcelado = true
    end
    
    def modificar_saldo(cantidad)
      @saldo = @saldo + cantidad
      @saldo
    end
    
    def obtener_capital
      valor_propiedades = 0
      @propiedades.each do |p|
        valor_propiedades = valor_propiedades + p.precio_compra +
                             p.num_casas * p.precio_edificar +
                             4 * p.num_hoteles * p.precio_edificar
        if(p.hipotecada)
          valor_propiedades = valor_propiedades - p.hipoteca_base
        end
      end
      valor_propiedades + @saldo
    end
    
    def obtener_propiedades(hipotecada)
      properties = Array.new
      @propiedades.each do |p|
        if(p.hipotecada == hipotecada)
            properties << p
        end
      end
      properties
    end
    
    def pagar_alquiler
      coste_alquiler = @casilla_actual.pagar_alquiler
      modificar_saldo(-coste_alquiler)
    end
    
    def pagar_impuesto
      coste_impuesto = @casilla_actual.coste 
      modificar_saldo(-coste_impuesto)
    end
    
    def pagar_libertad(cantidad)
      tengo_saldo = tengo_saldo(cantidad)
      if(tengo_saldo)
          @encarcelado = false
          modificar_saldo(-cantidad)
      end
    end
    
    def puedo_edificar_casa(titulo)
      num_casas = titulo.num_casas
      ((num_casas < 4) && es_de_mi_propiedad(titulo))
    end
    
    def puedo_edificar_hotel(titulo)
      num_casas = titulo.num_casas
      num_hoteles = titulo.num_hoteles
      (num_casas == 4 && num_hoteles < 4 && es_de_mi_propiedad(titulo))
    end
    
    def tengo_carta_libertad
      @tiene_carta_libertad 
    end
    
    def tengo_saldo(cantidad)
      @saldo > cantidad
    end
    
    def vender_propiedad(casilla)
      resultado = false
      titulo = casilla.titulo
      if(titulo != nil && es_de_mi_propiedad(titulo))
          eliminar_de_mis_propiedades(titulo)
          precio_venta = titulo.calcular_precio_venta
          modificar_saldo(precio_venta)
          casilla.titulo = nil
          resultado = true
      end
      resultado
    end
    
    def to_s
      "Jugador{encarcelado=#{@texto}, nombre=#{@nombre}, saldo=#{@saldo}, carta_libertad=#{@carta_libertad}, propiedades=#{@propiedades.join}, casilla_actual=#{@casilla_actual}}"
    end
    private :eliminar_de_mis_propiedades, :es_de_mi_propiedad, :tengo_saldo
  end
end

File: Practicas/Qytetet/QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;
import java.util.ArrayList;

/**
 *
 * @author maxigang
 */
public class Jugador implements Comparable {
    
    private boolean encarcelado;
    private String nombre;
    private int saldo;
    private boolean tieneCartaLibertad;
    
    private Sorpresa cartaLibertad;
    private ArrayList<TituloPropiedad> propiedades;
    private Casilla casillaActual;
    
    public Jugador(String n, int s)
    {
        encarcelado = false;
        nombre = n;
        saldo = s;
        tieneCartaLibertad = false;
        propiedades = new ArrayList<>();
    }
  
    protected Jugador(Jugador otroJugador)
    {
        encarcelado = otroJugador.encarcelado;
        nombre = otroJugador.nombre;
        saldo = otroJugador.saldo;
        tieneCartaLibertad = otroJugador.tieneCartaLibertad;
        cartaLibertad = otroJugador.cartaLibertad;
        propiedades = otroJugador.propiedades;
        casillaActual = otroJugador.casillaActual;
    }
  
    @Override
    public int compareTo(Object otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return otroCapital - obtenerCapital();
    }
    
    boolean cancelarHipoteca(TituloPropiedad titulo){
        int costeCancelarHipoteca = titulo.calcularCosteCancelar();
        boolean cancelada = false;
        if(tengoSaldo(costeCancelarHipoteca) && esDeMiPropiedadHipotecada(titulo,true))
        {
            modificarSaldo(-costeCancelarHipoteca);
            titulo.cancelarHipoteca();
            cancelada = true;
        }
        return cancelada;
    }
    
    boolean comprarTituloPropiedad(){
        boolean comprado = false;
        Calle calleActual = (Calle) casillaActual;
        int costeCompra = calleActual.getCoste();
        if(tengoSaldo(costeCompra))
        {
            TituloPropiedad titulo = calleActual.asignarPropietario(this);
            comprado = true;
            propiedades.add(titulo);
            modificarSaldo(-costeCompra);
        }
        return comprado;
    }
    
    ////
    
    protected Especulador convertirme(int fianza)
    {
        Especulador esp = new Especulador(this, fianza);
        return esp;
    }
    
    ////
    
    int cuantasCasasHotelesTengo(){
        int contador = 0;
        for(TituloPropiedad p:propiedades)
            contador += p.getNumCasas() + p.getNumHoteles();
        return contador;
    }
    
    ////
    
    protected boolean deboIrACarcel()
    {
        return !tengoCartaLibertad();
    }
    
    ////
    
    boolean deboPagarAlquiler(){
        Calle calleActual = (Calle) casillaActual;
        
        boolean esDeMiPropiedad = esDeMiPropiedad(calleActual.getTitulo());
        boolean tienePropietario = false;
        boolean estaEncarcelado = false;
        boolean estaHipotecada = false;
        
        if(!esDeMiPropiedad)
        {
            tienePropietario = calleActual.tengoPropietario();
            if(tienePropietario)
            {
                estaEncarcelado = calleActual.propietarioEncarcelado();
                estaHipotecada = calleActual.getTitulo().getHipotecada();
            }       
        }
        boolean deboPagar = !esDeMiPropiedad & tienePropietario & !estaEncarcelado & !estaHipotecada;
        
        return deboPagar;
    }
    
    Sorpresa devolverCartaLibertad(){ // RECORDAR QUE AL LLAMAR A ESTE METODO DEBEMOS ASEGURARNOS DE QUE EL JUGADOR TENGA UNA CARTA
        tieneCartaLibertad = false;
        return cartaLibertad;
    }
    
    boolean edificarCasa(TituloPropiedad titulo){
        boolean edificada = false;
        if(puedoEdificarCasa(titulo))
        {
            int costeEdificarCasa = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarCasa);
            if(tengoSaldo)
            {
                titulo.edificarCasa();
                modificarSaldo(-costeEdificarCasa);
                edificada = true;
            }
        }
        return edificada;
    }
    
    boolean edificarHotel(TituloPropiedad titulo){
        boolean edificada = false;
        if(puedoEdificarHotel(titulo))
        {
            int costeEdificarHotel = titulo.getPrecioEdificar();
            boolean tengoSaldo = tengoSaldo(costeEdificarHotel);
            if(tengoSaldo)
            {
                titulo.edificarHotel();
                modificarSaldo(-costeEdificarHotel);
                edificada = true;
            }
        }
        return edificada;
    }
    
    private void eliminarDeMisPropiedades(TituloPropiedad titulo){
        propiedades.remove(titulo);
        titulo.setPropietario(null);
    }
    
    protected boolean esDeMiPropiedad(TituloPropiedad titulo){
        for(TituloPropiedad p:propiedades)
            if(p == titulo)
                return true;
        return false;
    }
    
    boolean esDeMiPropiedadHipotecada(TituloPropiedad propiedad, boolean hipotecada)
    {
        for(TituloPropiedad p:obtenerPropiedades(hipotecada))
            if(p==propiedad)
                return true;
        return false;
    }
    
    /*boolean estoyEnCalleLibre(){ // NO SE LE DA USO
        throw new UnsupportedOperationException("Sin implementar");
    }*/
    
    Sorpresa getCartaLibertad(){
        return cartaLibertad;
    }
    
    public Casilla getCasillaActual(){
        return casillaActual;
    }
    
    boolean getEncarcelado(){
        return encarcelado;
    }
    
    ////
    
    protected int getFactorEspeculador()
    {
        return 0;
    }
    
    ////
    
    String getNombre(){
        return nombre;
    }
    
    ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    public int getSaldo(){
        return saldo;
    }
        
    boolean hipotecarPropiedad(TituloPropiedad titulo){
        boolean hipotecada = false;
        if(esDeMiPropiedadHipotecada(titulo,false))
        {
            int costeHipoteca = titulo.hipotecar();
            modificarSaldo(costeHipoteca);
            hipotecada = true;
        }
        return hipotecada;
    }
    
    void irACarcel(Casilla casilla){
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    int modificarSaldo(int cantidad){
        saldo += cantidad;
        return saldo;
    }
    
    int obtenerCapital(){
        int valor_propiedades = 0;
        for(TituloPropiedad p:propiedades)
        {
            valor_propiedades += p.getPrecioCompra() +
                                 p.getNumCasas() * p.getPrecioEdificar() +
                                 4 * p.getNumHoteles() * p.getPrecioEdificar();
            if(p.getHipotecada())
                valor_propiedades -= p.getHipotecaBase();
        }
        
        return valor_propiedades + saldo; 
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedades(boolean hipotecada){
        ArrayList<TituloPropiedad> properties = new ArrayList<>();
        for(TituloPropiedad p:propiedades)
            if(p.getHipotecada()==hipotecada)
                properties.add(p);
        return properties;
    }
    
    void pagarAlquiler(){
        Calle calleActual = (Calle) casillaActual;
        int costeAlquiler = calleActual.pagarAlquiler();
        modificarSaldo(-costeAlquiler);
    }
    
    protected void pagarImpuesto(){
        int costeImpuesto = casillaActual.getCoste(); 
        modificarSaldo(-costeImpuesto);
    }
    
    void pagarLibertad(int cantidad){
        boolean tengoSaldo = tengoSaldo(cantidad);
        if(tengoSaldo)
        {
            setEncarcelado(false);
            modificarSaldo(-cantidad);
        }
    }
    
    ////
    
    protected boolean puedoEdificarCasa(TituloPropiedad titulo)
    {
        int numCasas = titulo.getNumCasas();
        return ((numCasas < 4) && esDeMiPropiedad(titulo));
    }
    
    protected boolean puedoEdificarHotel(TituloPropiedad titulo)
    {
        int numCasas = titulo.getNumCasas();
        int numHoteles = titulo.getNumHoteles();
        return (numCasas == 4 && numHoteles < 4 && esDeMiPropiedad(titulo));
    }
    
    ////
    
    void setCartaLibertad(Sorpresa carta){
        cartaLibertad = carta;
        tieneCartaLibertad = true;
    }
    
    void setCasillaActual(Casilla casilla){
        casillaActual = casilla;
    }
    
    void setEncarcelado(boolean encarcelado){
        this.encarcelado = encarcelado;
    }
    
    boolean tengoCartaLibertad(){
        return tieneCartaLibertad;
    }
    
    protected boolean tengoSaldo(int cantidad){
        return saldo > cantidad;
    }
    
    boolean venderPropiedad(Calle casilla)
    {
        TituloPropiedad titulo = casilla.getTitulo();
        if(titulo != null && esDeMiPropiedad(titulo))
        {
            eliminarDeMisPropiedades(titulo);
            int precioVenta = titulo.calcularPrecioVenta();
            modificarSaldo(precioVenta);
            casilla.setTitulo(null);
            return true;
        }
        return false;
    }

    @Override
    public String toString() {
        return "Jugador{" + "encarcelado=" + encarcelado + ", nombre=" + nombre + 
               ", saldo=" + saldo + ", capital=" + obtenerCapital() + 
               ", cartaLibertad=" + cartaLibertad +
               ", propiedades=" + propiedades +
               ", casillaActual=" + casillaActual + '}';
    }
}